NOISSUE - Obtain email host from env

WashingtonKK opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

<!--

Pull request title should be `SMQ-XXX - description` or `NOISSUE - description` where XXX is ID of the issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/supermq/blob/main/CONTRIBUTING.md) file for detailed contributing guidelines.

For Work In Progress Pull Requests, please use the Draft PR feature, see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments.

- Provide tests for your changes.
- Use descriptive commit messages.
- Comment your code where appropriate.
- Squash your commits
- Update any related documentation.
-->

# What type of PR is this?

<!--This represents the type of PR you are submitting.

For example:
This is a bug fix because it fixes the following issue: #1234
This is a feature because it adds the following functionality: ...
This is a refactor because it changes the following functionality: ...
This is a documentation update because it updates the following documentation: ...
This is a dependency update because it updates the following dependencies: ...
This is an optimization because it improves the following functionality: ...
-->

## What does this do?

<!--
Please provide a brief description of what this PR is intended to do.
Include List any changes that modify/break current functionality.
-->

## Which issue(s) does this PR fix/relate to?

<!--
For pull requests that relate or close an issue, please include them below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "Resolves #1234" would connect the current pull request to issue 1234.  And when we merge the pull request, Github will automatically close the issue.
-->

- Related Issue #
- Resolves #

## Have you included tests for your changes?

<!--If you have not included tests, please explain why.
For example:
Yes, I have included tests for my changes.
No, I have not included tests because I do not know how to.
-->

## Did you document any new/modified feature?

<!--If you have not included documentation, please explain why.
For example:
Yes, I have updated the documentation for the new feature.
No, I have not updated the documentation because I do not know how to.
-->

### Notes

<!--Please provide any additional information you feel is important.-->

dborovcanin commented on pkg/sdk/sdk.go:
```diff
@@ -279,9 +279,9 @@ type SDK interface {
+	//  err := sdk.ResetPasswordRequest(context.Background(), "example@email.com", "https://example.com")
 	//  fmt.Println(err)
-	ResetPasswordRequest(ctx context.Context, email string) errors.SDKError
+	ResetPasswordRequest(ctx context.Context, email, host string) errors.SDKError
```
This should not be part of the SDK, but rather a configuration of the users service. I.e. you do not determine he host (URL, really) via (SDK) request.

SammyOina replied:
@dborovcanin should it be removed as part of the request? https://github.com/absmach/supermq/blob/e842bfaab6c359423d1544ee428b8a1a9bee4944/users/api/requests.go#L285

dborovcanin replied:
Yes.

dborovcanin reviewed (changes requested):
This should be solved on the Users service configuration, rather than in the SDK. Please address this problem there, and we can close this PR.

WashingtonKK committed:
remove referrer

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
update email

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
add host to env'

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
remove host from request body

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on docker/.env:
```diff
@@ -263,6 +263,7 @@ SMQ_OAUTH_UI_REDIRECT_URL=http://localhost:9095${SMQ_UI_PATH_PREFIX}/tokens/secu
 SMQ_OAUTH_UI_ERROR_URL=http://localhost:9095${SMQ_UI_PATH_PREFIX}/error
 SMQ_USERS_DELETE_INTERVAL=24h
 SMQ_USERS_DELETE_AFTER=720h
+SMQ_EMAIL_HOST_URL=http://localhost:9001
```
Let's rename to `SMQ_PASSWORD_RESET_URL_PREFIX`.

WashingtonKK committed:
update env

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on internal/email/email.go:
```diff
@@ -35,13 +35,14 @@ type email struct {
+	FromAddress string `env:"SMQ_EMAIL_FROM_ADDRESS"            envDefault:""`
+	FromName    string `env:"SMQ_EMAIL_FROM_NAME"               envDefault:""`
+	Template    string `env:"SMQ_EMAIL_TEMPLATE"                envDefault:"email.tmpl"`
+	HostURL     string `env:"SMQ_PASSWORD_RESET_URL_PREFIX"     envDefault:"http://localhost:8080"`
```
What's this used for?

dborovcanin commented on docker/docker-compose.yaml:
```diff
@@ -882,6 +882,7 @@ services:
       SMQ_SPICEDB_PRE_SHARED_KEY: ${SMQ_SPICEDB_PRE_SHARED_KEY}
       SMQ_SPICEDB_HOST: ${SMQ_SPICEDB_HOST}
       SMQ_SPICEDB_PORT: ${SMQ_SPICEDB_PORT}
+      SMQ_PASSWORD_RESET_URL_PREFIX: ${SMQ_PASSWORD_RESET_URL_PREFIX}
```
Let's add `USERS` also - `SMQ_USERS_PASSWORD_RESET_URL_PREFIX`.

WashingtonKK committed:
update url name

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on internal/email/email.go:
```diff
@@ -35,13 +35,14 @@ type email struct {
+	FromAddress                 string `env:"SMQ_EMAIL_FROM_ADDRESS"             envDefault:""`
+	FromName                    string `env:"SMQ_EMAIL_FROM_NAME"                envDefault:""`
+	Template                    string `env:"SMQ_EMAIL_TEMPLATE"                 envDefault:"email.tmpl"`
+	UsersPasswordResetUrlPrefix string `env:"SMQ_USERS_PASSWORD_RESET_URL_PREFIX" envDefault:"http://localhost:8080"`
```
No need to map field name with env; simplify to `PasswordResetURLPrefix`.

dborovcanin commented on users/events/streams.go:
```diff
@@ -304,15 +304,14 @@ func (es *eventStore) Identify(ctx context.Context, session authn.Session) (stri
 
-func (es *eventStore) GenerateResetToken(ctx context.Context, email, host string) error {
-	err := es.svc.GenerateResetToken(ctx, email, host)
+func (es *eventStore) GenerateResetToken(ctx context.Context, email string) error {
```
This name is misleading. It does not generate the token, but also sends an email. Consequentially, `SendPasswordReset` is useless. Let's remove it and rename `GenerateResetToken` to `SendPasswordReset`.

WashingtonKK committed:
refactor reset password

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
update endpoint tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on internal/email/email.go:
```diff
@@ -35,13 +35,14 @@ type email struct {
+	FromAddress            string `env:"SMQ_EMAIL_FROM_ADDRESS"              envDefault:""`
+	FromName               string `env:"SMQ_EMAIL_FROM_NAME"                 envDefault:""`
+	Template               string `env:"SMQ_EMAIL_TEMPLATE"                  envDefault:"email.tmpl"`
+	PasswordResetURLPrefix string `env:"SMQ_USERS_PASSWORD_RESET_URL_PREFIX" envDefault:"http://localhost:8080"`
```
Don't pollute `internal/email` with this. Email has nothing to do with password reset; it only send whatever email you get him send. Rather, do something like (i.e. what we already have, but don't pass host param to the `SendPasswordReset`:

```suggestion

type emailer struct {
	resetURL string
	agent    *email.Agent
}

// New creates new emailer utility.
func New(resetURL string, c *email.Config) (users.Emailer, error) {
	e, err := email.New(c)
	return &emailer{
		resetURL: resetURL,
		agent:    e,
	}, err
}

func (e *emailer) SendPasswordReset(to []string, user, token string) error {
	cont := fmt.Sprintf("%s?token=%s", e.resetURL, token)
	return e.agent.Send(to, "", "Password Reset Request", "", user, url, "")
}

```

WashingtonKK committed:
update emailer

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
lint

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on internal/email/email.go:
```diff
@@ -35,13 +35,13 @@ type email struct {
+	Password    string `env:"SMQ_EMAIL_PASSWORD"                  envDefault:""`
+	FromAddress string `env:"SMQ_EMAIL_FROM_ADDRESS"              envDefault:""`
+	FromName    string `env:"SMQ_EMAIL_FROM_NAME"                 envDefault:""`
+	Template    string `env:"SMQ_EMAIL_TEMPLATE"                  envDefault:"email.tmpl"`
```
There's no need for this change. Let's keep it one whitespace from the longest-named field.

dborovcanin commented on users/users.go:
```diff
@@ -179,9 +179,8 @@ type Service interface {
-	// host is used for generating reset link.
-	GenerateResetToken(ctx context.Context, email, host string) error
+	// SendPasswordReset sends reset password link to email.
+	SendPasswordReset(ctx context.Context, email string) error
```
```
// SendPasswordReset generates reset password link and sends it to the user via email.
```

dborovcanin commented on cmd/users/main.go:
```diff
@@ -67,28 +67,29 @@ const (
+	SpicedbHost            string        `env:"SMQ_SPICEDB_HOST"              envDefault:"localhost"`
+	SpicedbPort            string        `env:"SMQ_SPICEDB_PORT"              envDefault:"50051"`
+	SpicedbPreSharedKey    string        `env:"SMQ_SPICEDB_PRE_SHARED_KEY"    envDefault:"12345678"`
+	PasswordResetURLPrefix string        `env:"SMQ_USERS_PASSWORD_RESET_URL_PREFIX" envDefault:"http://localhost:8080"`
```
Same here, align tags to the longest-named field.

WashingtonKK committed:
style(email): align Config env tags; docs(users): clarify password reset

- Align env tag spacing in internal/email/email.go for consistency and readability.
- Clarify SendPasswordReset comment in users/users.go to mention link generation and email delivery.

No functional changes.

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
style: normalize Go struct formatting in users and email

- Convert space-indented struct fields to Go-standard tabs.
- Align env tag columns for readability and consistency.

Files:
- cmd/users/main.go (config struct)
- internal/email/email.go (Config struct)

No functional changes.

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on cmd/users/main.go:
```diff
@@ -67,28 +67,29 @@ const (
+	AdminFirstName         string        `env:"SMQ_USERS_ADMIN_FIRST_NAME"          envDefault:"super"`
+	AdminLastName          string        `env:"SMQ_USERS_ADMIN_LAST_NAME"           envDefault:"admin"`
+	PassRegexText          string        `env:"SMQ_USERS_PASS_REGEX"                envDefault:"^.{8,}$"`
+	ResetURL               string        `env:"SMQ_TOKEN_RESET_ENDPOINT"            envDefault:"/reset-request"`
```
No need for both this and `PasswordResetURLPrefix`. Let's only use the `PasswordResetURLPrefix`.

WashingtonKK committed:
fix(users): use /reset-request for password reset

Remove SMQ_TOKEN_RESET_ENDPOINT from users service config and build the emailer URL as {PasswordResetURLPrefix}/reset-request to align with the expected route.

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
chore(users): remove unused SMQ_TOKEN_RESET_ENDPOINT env

Drop SMQ_TOKEN_RESET_ENDPOINT from docker/.env, docker-compose.yaml, and users/README.md.\nThe users service now builds the reset link as {PasswordResetURLPrefix}/reset-request, making this env redundant.

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
chore(users): rename password reset URL env

Rename SMQ_USERS_PASSWORD_RESET_URL_PREFIX to SMQ_PASSWORD_RESET_URL_PREFIX across code and Docker configs.\nThis aligns naming with the users service config and follows removal of the redundant SMQ_TOKEN_RESET_ENDPOINT, keeping a single source for building reset links ({PasswordResetURLPrefix}/reset-request).

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>
